### ISSUE TITLE
🚀 Feature: Digest/Wait Until

### ISSUE THREAD

**Issue by paulwer** [notifirehq/notifire#6915]
### 🔖 Feature description

We would like to see a callback function to check if a condition is true to f.ex. end a digest or wait step.<br>If leveraged, this could also lead to be used as schedules for checking workflows to be run. But this may be out of scope.

### 🎤 Why is this feature needed ?

This would enable followup emails or even responses to interaction more smoothly and would reduce the developers need to setup complex cronjobs for this.

Use-Cases I can think of:

* Error was resolved
* Errors & Warnings Workflow detected an error with a high priority, so it will skip the current digest and will populate all current digest events and the current event
* Thanks for your reply f.ex. in discord => maybe such use-cases can get much more complex to be honest :)

### ✌️ How do you aim to achieve this?

1. condition check function => the function which will check if a condition was met
2. ttl or trigger event-selector => define how often this should be triggered
3. BackOff strategy => to reduce systemload after a while

### 🔄️ Additional Information

*No response*

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](<https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md>)

### Are you willing to submit PR?

None

**Comment by Abobos** [comment]
@scopsy can I take a look at this issue?

**Comment by paulwer** [comment]
This could also be extened with a triggerEvent function within the sdk to trigger those waitUntil steps manualy to be completed with the user backend as an initiator. Maybe an optional maxTimeout could be usefull.

Additional use case:
- having a long running job triggering a workflow for a subscriber for multiple things within the process and proceed with all workflows to send emails, when the long-running job end, ensuring sending only one email.

**Comment by paulwer** [comment]
@scopsy we just encountered a use-case for this again, so I wanted to reask, if there is still a probability for this getting looked at / implemented?

I think this feature can be designed without a breaking change, by making all schedule related properties optional and let the user create the step via the framework as usual. (i mean an extension of the digest functionality instead of creating a new step type)

Triggering the digest to complete would also allow the frontend to display a skip digest button for workflows waiting for the digest to complete, which would be very usefull for testing and running workflows via the dashboard.

When it has schedule configuration like cron or lookback etc, it would remove the scheduled job and complete the digest early and when there is no scheduled job it would complete the digest manualy.

**Comment by scopsy** [comment]
Thanks @paulwer, unfortuently not at the moment :( We recently introduced a new HTTP Step type, which allows to fetch specific data, but it won't allow finishing a digest or continuing the flow as you might need.

I think that the besgt approach would be to introduced a new Step type for this, as it will create the most flexibility. 

So you would imaging it work in a way that if a wait until step is met, there will be a unique url or re-trigger that will continue the workflow to the next state?

**Comment by paulwer** [comment]
@scopsy I do think, it would be smarter to extend existing digest step/api, because it does not introduce breaking changes and would allow users to use it more flexible and existing frontend can be easlily reused/extended.

I would suggest the following changes: (which can be combined freely)

1. CRUD the digest:
    from my pov we currently have like 2 trigger-types: with offset/lookback & cron, you would add the ability to define no trigger, which basicly would act as a wait until

2. add 2 workflow/transaction related actions
   a) Complete a Digest
       Could be used in 2 way:
           - early complete a digest which waits on the trigger condition (cron/offset) to be met
           - complete the wait until
   b) Cancel a Digest/Transaction
       Eigher change the cancel transaction endpoint to support a query. f.ex.
           - transactionId
           - subscriberId
           - digestId
           - topicName
           - etc....
        This would allow the users to dynamicly cancel multiple workflows/digest
         f.ex. to met requirements to cancel sending email to a subscriber, which was deletet/removed from an organization or simmilar

In the frontend you can the actions apropiatly as well :)

**Comment by paulwer** [comment]
@scopsy I drafted a PR and will verify/test it shortly. please feel free to give me feedback if that could/would be included into the platform.
